cmd/crank/beta/validate/validate.go

```diff
@@ -0,0 +1,166 @@
+			xrd := &v1.CompositeResourceDefinition{}
+			bytes, err := s.MarshalJSON()
+			if err != nil {
+				return nil, errors.Wrap(err, "cannot marshal XRD to JSON")
+			}
+
+			if err := yaml.Unmarshal(bytes, xrd); err != nil {
+				return nil, errors.Wrap(err, "cannot unmarshal XRD YAML")
+			}
+
+			crd, err := xcrd.ForCompositeResource(xrd)
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot derive composite CRD from XRD %q", xrd.GetName())
+			}
+			crds = append(crds, crd)
+
+			if xrd.Spec.ClaimNames != nil {
+				claimCrd, err := xcrd.ForCompositeResourceClaim(xrd)
+				if err != nil {
+					return nil, errors.Wrapf(err, "cannot derive claim CRD from XRD %q", xrd.GetName())
+				}
+				crds = append(crds, claimCrd)
+			}
+
+		case schema.GroupKind{Group: "pkg.crossplane.io", Kind: "Provider"}:
+			fmt.Println("Provider extension is not supported yet")
+			continue
+
+		case schema.GroupKind{Group: "pkg.crossplane.io", Kind: "Configuration"}:
+			fmt.Println("Configuration extension is not supported yet")
+			continue
+
+		default:
+			continue
+		}
+	}
+
+	return crds, nil
+}
+
+func newValidators(crds []*extv1.CustomResourceDefinition) (map[schema.GroupVersionKind]validation.SchemaValidator, error) {
+	validators := map[schema.GroupVersionKind]validation.SchemaValidator{}
+
+	for i := range crds {
+		internal := &ext.CustomResourceDefinition{}
+		if err := extv1.Convert_v1_CustomResourceDefinition_To_apiextensions_CustomResourceDefinition(crds[i], internal, nil); err != nil {
+			return nil, err
+		}
+
+		for _, ver := range internal.Spec.Versions {
+			var sv validation.SchemaValidator
+			var err error
+
+			if ver.Schema != nil && ver.Schema.OpenAPIV3Schema != nil {
+				sv, _, err = validation.NewSchemaValidator(ver.Schema.OpenAPIV3Schema)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			if internal.Spec.Validation != nil {
+				sv, _, err = validation.NewSchemaValidator(internal.Spec.Validation.OpenAPIV3Schema)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			validators[schema.GroupVersionKind{
+				Group:   internal.Spec.Group,
+				Version: ver.Name,
+				Kind:    internal.Spec.Names.Kind,
+			}] = sv
+		}
+	}
+
+	return validators, nil
+}
+
+func validateResources(resources []*unstructured.Unstructured, crds []*extv1.CustomResourceDefinition, skipSuccessLogs bool) error {
+	validators, err := newValidators(crds)
+	if err != nil {
+		return errors.Wrap(err, "cannot create validators")
+	}
+
+	failure, warning := 0, 0
+
+	for i, r := range resources {
+		v, ok := validators[r.GetObjectKind().GroupVersionKind()]
+		if !ok {
+			warning++
+			fmt.Println("[!] could not find CRD/XRD for: " + r.GroupVersionKind().String())
+			continue
+		}
+
+		re := v.Validate(&resources[i])
+		for _, e := range re.Errors {
+			failure++
+			fmt.Printf("[x] validation error %s - %s : %s\n", r.GroupVersionKind().String(), r.GetAnnotations()[composite.AnnotationKeyCompositionResourceName], e.Error())
```

Comment:
i notice i'm having a bit of trouble reading the output when the group/version, the kind, and the name are all split into different fields.

I was trying to suggest something better, i.e. a way that unifies them into a single string that may be easier to read, but I don't think that's really a thing in Kubernetes when you are including the entire GVK as well as the resource name, and we are in fact using the official `GVK.String()` function - so I can't complain much 😂 

Perhaps something as simple as using a `,` instead of `-` before the name field could make it feel more consistent? Like this:

```
iam.aws.upbound.io/v1beta1, Kind=AccessKey, sample-access-key-0
```